tutorials/lesson3-anatomy/coin.template.qnt

```diff
@@ -0,0 +1,953 @@
+
+        </description>
+        <file>$file</file>
+        <line>$line</line>
+      </step>
+    !*/
+
+    // All possible behaviors of the protocol in one action.
+    // Note that random simulation produces valid inputs to
+    // this action with a low probability, since amount are chosen at random.
+    // A symbolic model checker such as Apalache does not have any problem
+    // with this. Random simulation requires a more restricted version of step.
+    action step: bool = {
+        nondet sender = oneOf(ADDR)
+        nondet receiver = oneOf(ADDR)
+        nondet amount = 0.to(MAX_UINT).oneOf()
+        // execute one of the available actions/methods
+        any {
+            mint(sender, receiver, amount),
+            send(sender, receiver, amount),
+        }
+    }
+    /*!
+      <step>
+        <title>Defining a protocol step</title>
+        <description>
+
+Finally, we can put together `mint` and `send` to describe every possible
+transition of the protocol! To this end, we non-deterministically choose
+the sender, the receiver, and the amount (from the set of integers from 0
+to `MAX_UINT`, inclusive). The expression `any { ... }` executes one of its
+arguments; if both could be executed, one of them is executed
+non-deterministically.
+
+**Exercise:** Run REPL, execute `init` once and execute `step` multiple times.
+Do you understand why some of the occurrences of `step` evaluate to `false`
+and some evaluate to `true`? It may help you, if you print `state` after `init`
+and `step`.
+
+We have defined the state variables, the initializer, and the protocol stop.
+This is the minimal set of prerequisites for defining a working protocol.
+What we have achieved here is great! You can play with the protocol, feed it
+with different parameters in REPL, and experiment with your protocol.
+
+Although you should definitely play with the protocol at this point, we urge you
+not to stop here! In the next steps, we will show you the real magic of Quint.
+        </description>
+        <file>$file</file>
+        <line>$line</line>
+      </step>
+    !*/
+
+    // INVARIANTS AND TEMPORAL PROPERTIES
+    /*!
+      <step>
+        <title>Expected properties</title>
+        <description>
+
+Having defined the protocol behavior with `init` and `step`, it is time to think
+about what we expect from the protocol. This is a good place for specifying
+protocol invariants and temporal properties.
+
+        </description>
+        <file>$file</file>
+        <line>$line</line>
+      </step>
+    !*/
+
+    // One of the simplest properties is that all balances are within the uint range.
+    // While it is automatically guaranteed by Solidity, our specification is using
+    // big integers. Hence, it makes sense to check the range.
+    val balancesRangeInv: bool = {
+        ADDR.forall(a =>
+            val b = balances.get(a)
+            (0 <= b) and (b <= MAX_UINT)
```

Comment:
Why not use `isUInt`?

Reply:
brilliant! This makes the invariant a one-liner!